File: source/data/mappers/src/jvmMain/kotlin/com/voxeldev/tgdrive/mappers/media/FileMapper.kt
package com.voxeldev.tgdrive.mappers.media

import com.voxeldev.tgdrive.main.api.files.FileLocal
import com.voxeldev.tgdrive.main.api.files.FileRemote
import org.drinkless.tdlib.TdApi.File
import org.drinkless.tdlib.TdApi.LocalFile
import org.drinkless.tdlib.TdApi.RemoteFile
import com.voxeldev.tgdrive.main.api.files.File as CommonFile

/**
 * @author nvoxel
 */
class FileMapper {

    fun toResponse(file: File): CommonFile =
        CommonFile(
            id = file.id,
            size = file.size,
            expectedSize = file.expectedSize,
            local = file.local.toResponse(),
            remote = file.remote.toResponse(),
        )

    private fun LocalFile.toResponse(): FileLocal =
        FileLocal(
            path = path,
            canBeDownloaded = canBeDownloaded,
            canBeDeleted = canBeDeleted,
            isDownloadingActive = isDownloadingActive,
            isDownloadingCompleted = isDownloadingCompleted,
            downloadOffset = downloadOffset,
            downloadedPrefixSize = downloadedPrefixSize,
            downloadedSize = downloadedSize,
        )

    private fun RemoteFile.toResponse(): FileRemote =
        FileRemote(
            id = id,
            uniqueId = uniqueId,
            isUploadingActive = isUploadingActive,
            isUploadingCompleted = isUploadingCompleted,
            uploadedSize = uploadedSize,
        )
}

File: iosApp/iosApp/data/mappers/media/FileMapper.swift
//
//  FileMapper.swift
//  iosApp
//
//  Created by nVoxel on 25.04.2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import Foundation
import TDLibKit
import TGDriveKit

class FileMapper {
    
    func toResponse(file: File) -> CommonFile {
        return CommonFile(
            id: Int32(file.id),
            size: file.size,
            expectedSize: file.expectedSize,
            local: toResponse(localFile: file.local),
            remote: toResponse(remoteFile: file.remote)
        )
    }
    
    private func toResponse(localFile: LocalFile) -> CommonFileLocal {
        return CommonFileLocal(
            path: localFile.path,
            canBeDownloaded: localFile.canBeDownloaded,
            canBeDeleted: localFile.canBeDeleted,
            isDownloadingActive: localFile.isDownloadingActive,
            isDownloadingCompleted: localFile.isDownloadingCompleted,
            downloadOffset: localFile.downloadOffset,
            downloadedPrefixSize: localFile.downloadedPrefixSize,
            downloadedSize: localFile.downloadedSize
        )
    }
    
    private func toResponse(remoteFile: RemoteFile) -> CommonFileRemote {
        return CommonFileRemote(
            id: remoteFile.id,
            uniqueId: remoteFile.uniqueId,
            isUploadingActive: remoteFile.isUploadingActive,
            isUploadingCompleted: remoteFile.isUploadingCompleted,
            uploadedSize: remoteFile.uploadedSize
        )
    }
}
